### PR TITLE
[cmd/mdatagen] Add unit test for metric collisions.

### PR DESCRIPTION
**Pull request by mackjmr**
**Description:**
This PR adds a unit test to check for metric collisions in receivers which have metrics defined.

**Link to tracking Issue:**
#23375

### CODE REVIEW

**Review by songy23 — APPROVED** [review]

File: cmd/mdatagen/validate_test.go [hunk 1]
```diff
@@ -101,3 +105,38 @@ func TestValidate(t *testing.T) {
...
+	seen := make(map[string]string)
+	for receiver, metrics := range allMetrics {
+		for _, metricName := range metrics {
+			// TODO: Remove one https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/24009 and
+			// https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/24008 have been addressed.
+			if metricName == "container.cpu.utilization" || metricName == "container.memory.rss" {
+				continue
+			}
+			val, ok := seen[metricName]
+			assert.False(t, ok, fmt.Sprintf("Collision for metric %v in receivers %v and %v \n", metricName, receiver, val))
```
Comment (crobert-1) [inline cmd/mdatagen/validate_test.go:138]: Feel free to disregard, I understand that `ok` may be a bit of the default name here for whether a key exists in the map, but it seems counter-intuitive in this case. A test passing because something is not `ok` is just a bit hard to follow. 

Maybe `exists` or `previouslySeen`, something like that? Not a big deal though.

**Review by crobert-1 — APPROVED** [review]

File: cmd/mdatagen/validate_test.go (hunk 1 above)
Comment (mackjmr) [reply cmd/mdatagen/validate_test.go]: Agreed, changed `ok` to `exists` ✔️

**Comment by crobert-1** [comment]
CI [action failure](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/5540434478/jobs/10112716619?pr=24229) is frequency of #24223.

**Review by atoulme — APPROVED** [review]

File: cmd/mdatagen/validate_test.go [hunk 2]
```diff
@@ -101,3 +105,38 @@ func TestValidate(t *testing.T) {
 		})
 	}
 }
+
+func TestMetricCollision(t *testing.T) {
+	allMetrics := map[string][]string{}
+	err := filepath.Walk("../../receiver", func(path string, info fs.FileInfo, err error) error {
+		if info.Name() == "metadata.yaml" {
+			// TODO: Remove once https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24011 is merged.
```
Comment (dmitryax) [inline cmd/mdatagen/validate_test.go:113]: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24011 is merged. Please rebase

File: cmd/mdatagen/validate_test.go [hunk 3]
```diff
@@ -101,3 +105,34 @@ func TestValidate(t *testing.T) {
...
+			// TODO: Remove one https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/24009 and
+			// https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/24008 have been addressed.
+			if metricName == "container.cpu.utilization" || metricName == "container.memory.rss" {
+				continue
+			}
+			val, exists := seen[metricName]
+			assert.False(t, exists, fmt.Sprintf("Collision for metric %v in receivers %v and %v \n", metricName, receiver, val))
+			seen[metricName] = receiver
+		}
+	}
```
Comment (dmitryax) [inline cmd/mdatagen/validate_test.go:137]: I don't think we ever decided that each receiver can emit unique metric names. I'd say we should remove this section until this decision is made.

**Review by dmitryax — CHANGES_REQUESTED** [review]
I just realized that this whole test is about making sure that each receiver emits unique metrics which we don't have a rule for. I think we shouldn't introduce it until we decide on this rule

**Comment by mackjmr** [comment]
Opened an issue: #26499.
